xdsl/transforms/csl_stencil_set_global_coeffs.py

```diff
@@ -0,0 +1,189 @@
+    """
+    d, max_distance = get_dir_and_distance(op.offset)
+    return csl.DirectionOp(d), arith.ConstantOp(IntegerAttr(max_distance, 16))
+
+
+def get_coeff_api_ops(op: csl_stencil.ApplyOp, wrapper: csl_wrapper.ModuleOp):
+    coeffs = list(op.coeffs or [])
+    elem_t = Float32Type()
+    pattern = wrapper.get_param_value("pattern").value.data
+    neighbours = pattern - 1
+    is_wse2 = wrapper.target.data == "wse2"
+    if is_wse2:
+        empty: list[FloatAttr] = [FloatAttr(f, elem_t) for f in [0] + neighbours * [1]]
+        shape = (pattern,)
+    else:
+        empty: list[FloatAttr] = [FloatAttr(f, elem_t) for f in neighbours * [1]]
+        shape = (pattern - 1,)
+
+    cmap: dict[csl.Direction, list[FloatAttr]] = {
+        csl.Direction.NORTH: empty,
+        csl.Direction.SOUTH: empty.copy(),
+        csl.Direction.EAST: empty.copy(),
+        csl.Direction.WEST: empty.copy(),
+    }
+
+    for c in coeffs:
+        direction, distance = get_dir_and_distance(c.offset)
+        if not is_wse2:
+            distance -= 1
+        cmap[direction][distance] = c.coeff
+
+    memref_t = memref.MemRefType(elem_t, shape)
+    ptr_t = csl.PtrType.get(memref_t, is_single=True, is_const=True)
+
+    cnsts = {
+        d: arith.ConstantOp(DenseIntOrFPElementsAttr.create_dense_float(memref_t, v))
+        for d, v in cmap.items()
+    }
+    addrs = {d: csl.AddressOfOp(v, ptr_t) for d, v in cnsts.items()}
+
+    # pretty-printing
+    for d, c in cnsts.items():
+        c.result.name_hint = str(d)
+
+    args: list[Operation] = [
+        addrs[csl.Direction.EAST],
+        addrs[csl.Direction.WEST],
+        addrs[csl.Direction.SOUTH],
+        addrs[csl.Direction.NORTH],
+    ]
+
+    return [
+        *cnsts.values(),
+        *args,
+        csl.MemberCallOp(
+            "setCoeffs",
+            None,
+            wrapper.get_program_import("stencil_comms.csl"),
+            args,
+        ),
+    ]
+
+
+@dataclass(frozen=True)
+class GenerateCoeffAPICalls(RewritePattern):
+    """
+    Generates calls to the stencil_comms API to set coefficients.
+
+    The API currently supports only f32 coeffs.
+    """
+
+    @op_type_rewrite_pattern
+    def match_and_rewrite(self, op: csl_wrapper.ModuleOp, rewriter: PatternRewriter, /):
+        applies: list[csl_stencil.ApplyOp] = []
+        global_coeffs = []
+        # first_apply = None
+        for apply in op.walk():
+            if isinstance(apply, csl_stencil.ApplyOp):
+                if not applies:
+                    if apply.coeffs:
+                        global_coeffs = sorted(
+                            apply.coeffs.data, key=lambda x: x.offset
+                        )
+                elif global_coeffs != (
+                    sorted(apply.coeffs.data, key=lambda x: x.offset)
+                    if apply.coeffs
+                    else []
+                ):
+                    return
+                applies.append(apply)
```

Comment:
This was quite confusing on first parse. Not sure how it can be improved much, a comment would help though